drivers/axi_core/spi_engine/spi_engine.c

```diff
@@ -0,0 +1,943 @@
+
+	desc_extra = cast_to_extra_desc(desc->extra);
+
+	command = (cmd & cmd_msk);
+	param = (cmd & param_msk);
+
+	switch(command) {
+	case CS_DEASSERT:
+		/* Reet the CS */
+		spi_eng_gen_cs(desc, xfer, false);
+		break;
+
+	case CS_ASSERT:
+		/* Set the CS */
+		spi_eng_gen_cs(desc, xfer, true);
+		break;
+
+	case SLEEP_CMD:
+		/* Sleep */
+		spi_gen_sleep_ns(desc, xfer, param);
+		break;
+
+	case TRANSFER_R_CMD:
+		if(spi_check_dma_config(desc_extra, 1, 0) == SUCCESS) {
+			/* Read */
+			spi_eng_gen_transfer(desc_extra,
+					     xfer, false, true, param);
+			desc_extra->rx_length = param;
+		} else {
+			printf("%s: DMA Rx not configured.\n", __func__);
+			desc_extra->rx_length = 0;
+		}
+
+		break;
+
+	case TRANSFER_W_CMD:
+		if(spi_check_dma_config(desc_extra, 0, 1) == SUCCESS) {
+			/* Write */
+			spi_eng_gen_transfer(desc_extra,
+					     xfer, true, false, param);
+			desc_extra->tx_length = param;
+		} else {
+			printf("%s: DMA Tx not configured.\n", __func__);
+			desc_extra->tx_length = 0;
+		}
+		break;
+
+	case TRANSFER_R_W_CMD:
+		if(spi_check_dma_config(desc_extra, 1, 1) == SUCCESS) {
+			/* Read and write */
+			spi_eng_gen_transfer(desc_extra,
+					     xfer, true, true, param);
+			desc_extra->tx_length = param;
+			desc_extra->rx_length = param;
+		} else {
+			printf("%s: DMA Rx and Tx not configured.\n", __func__);
+			desc_extra->tx_length = 0;
+			desc_extra->rx_length = 0;
+		}
+		break;
+
+	default:
+		break;
+	}
+}
+
+/*******************************************************************************
+ *
+ * @func	spi_eng_compile_message
+ *
+ * @brief	Prepare the command queue before sending it to the engine
+ *
+ * @params
+ *		desc		- Spi engine descriptor
+ *		msg		- Structure used to store the messages
+ *		xfer		- Fifo transfer structure
+ *
+ * @return			- This function allways returns SUCCESS
+ *
+ ******************************************************************************/
+int32_t spi_eng_compile_message(spi_desc *desc,
+				spi_eng_msg *msg,
+				spi_eng_transfer_fifo *xfer)
+{
+	uint32_t	i;
+	uint32_t	n;
+	spi_desc_extra	*desc_extra;
+
+	desc_extra = cast_to_extra_desc(desc->extra);
+
+	n = msg->msg_cmd_len;
+
+	/* Configure the prescaler */
+	spi_eng_program_add_cmd(xfer,
+				SPI_ENGINE_CMD_WRITE(SPI_ENGINE_CMD_REG_CLK_DIV,
+						desc_extra->clk_div));
+	/*
+	 * Configure the spi mode :
+	 *	- 3 wire
+	 *	- CPOL
+	 *	- CPHA
+	 */
+	spi_eng_program_add_cmd(xfer,
+				SPI_ENGINE_CMD_WRITE(SPI_ENGINE_CMD_REG_CONFIG,
+						desc->mode));
+
+	/* Set the data transfer length */
+	spi_eng_program_add_cmd(xfer,
+				SPI_ENGINE_CMD_WRITE(SPI_ENGINE_CMD_DATA_TRANSFER_LEN,
+						desc_extra->data_width));
+
+	/* Add a sync command to signal that the transfer has started */
+	spi_eng_program_add_cmd(xfer,
+				SPI_ENGINE_CMD_SYNC(SPI_ENGINE_SYNC_TRANSFER_BEGIN));
+
+	/* Add the rest of the commands */
+	for (i = 0; i < n; i++) {
+		spi_eng_add_user_cmd(desc, xfer, msg->spi_msg_cmds[i]);
+	}
```

Comment:
these brackets can be removed. If you are doing this revision, a clean-up would not hurt.